### PR TITLE
Perf optimizations

### PR DESCRIPTION
**Pull request by divmain**
This PR is in support of https://github.com/FormidableLabs/victory-chart/pull/93.  It include one hot-path performance improvement, as well as a multi-arg memoization function that is consumed by `victory-chart`.

It looks like I have to update a couple of tests for `getStyles`.  Also, this PR will probably have to wait until after the `victory-core` merge.

/cc @boygirl @coopy

### CODE REVIEW

File: src/perf.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+export const memoize = function (fn) {
+  const cache = {};
+  return function () {
+    const args = Array.prototype.slice.call(arguments);
+    const hash = args.map((arg) => {
+      return (typeof arg === "string" || typeof arg === "number") ? arg : JSON.stringify(arg);
```
Comment (coopy) [inline src/perf.js:6]: This won't work for `function` args; it might be worth noting in a comment.

File: src/perf.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export const memoize = function (fn) {
```
Comment (ryan-roemer) [inline src/perf.js:1]: Can you paste a code comment with a reference / links to where we're getting our memoize function from? There are a ton of battle-tested examples out there, and hopefully we can leverage a good, well-trodden path here.

Comment (divmain) [inline src/perf.js:1]: Happy to make a change here if you think it is necessary, but let me explain my thinking.  If you look at the `memoize` function that is included with Lodash, there isn't much to it.  If a `resolver` is included, all arguments _will_ be passed (this is not the case with Underscore's `memoize`), which makes using `Lodash#memoize` a possibility.

However, the `resolver` key-generation will consist of basically all of the code in this PR's `memoize`, plus overhead of an additional `fn.apply`.  Since the motivation for this PR is to improve performance, and because this `memoize` is in the hot code path, I wanted to reduce any unnecessary overhead.

In other words, I'm fine changing this to `_.memoize(fn, /* all the code in perf.js */)`, but it doesn't seem to buy us anything and it will mean a slight perf hit.

File: src/perf.js (hunk 1 above)
Comment (divmain) [inline src/perf.js:6]: Good thought.

**Comment by divmain** [comment]
I'm moving this PR over to `victory-core`, will check back here in case you reply @ryan-roemer.
